File: src/lib/api/private-opportunities.ts
import { apiFetch, apiFetchArray, apiPost, apiPut, apiDelete } from "../api";

export interface PrivateOpportunity {
  id: string;
  user_id: string;
  type: "grant" | "scholarship";
  title: string;
  provider: string;
  description: string | null;
  eligibility_criteria: Record<string, unknown> | null;
  award_range: string | null;
  deadline: string | null;
  field_tags: string[] | null;
  region: string | null;
  source_url: string | null;
  guidelines: string | null;
  is_parsed: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface ParsedExternal {
  title: string | null;
  provider: string | null;
  description: string | null;
  deadline: string | null;
  award_range: string | null;
  eligibility_criteria: Record<string, unknown> | null;
  field_tags: string[] | null;
  region: string | null;
  source_url: string | null;
}

const USER_ID = "user-001";

export async function fetchPrivateOpportunities(type?: "grant" | "scholarship"): Promise<PrivateOpportunity[]> {
  const params: Record<string, string> = { user_id: USER_ID };
  if (type) params.type = type;
  return apiFetchArray<PrivateOpportunity>("/api/opportunities/private", { params });
}

export async function getPrivateOpportunity(id: string): Promise<PrivateOpportunity | null> {
  return apiFetch<PrivateOpportunity>(`/api/opportunities/private/${id}`);
}

export async function createPrivateOpportunity(data: {
  type: "grant" | "scholarship";
  title: string;
  provider: string;
  description?: string;
  eligibility_criteria?: Record<string, unknown>;
  award_range?: string;
  deadline?: string;
  field_tags?: string[];
  region?: string;
  source_url?: string;
  guidelines?: string;
  is_parsed?: boolean;
}): Promise<PrivateOpportunity | null> {
  return apiPost<PrivateOpportunity>("/api/opportunities/private", {
    ...data,
    user_id: USER_ID,
  });
}

export async function updatePrivateOpportunity(
  id: string,
  data: Partial<Omit<PrivateOpportunity, "id" | "user_id" | "created_at" | "updated_at">>
): Promise<PrivateOpportunity | null> {
  return apiPut<PrivateOpportunity>(`/api/opportunities/private/${id}`, data);
}

export async function deletePrivateOpportunity(id: string): Promise<boolean> {
  return apiDelete(`/api/opportunities/private/${id}`);
}

export async function parseExternalUrl(url: string, type: "grant" | "scholarship" = "grant"): Promise<ParsedExternal | null> {
  return apiPost<ParsedExternal>("/api/opportunities/parse-external", { url, type });
}

export async function createApplicationFromOpportunity(
  opportunityId: string,
  type: "grant" | "scholarship"
): Promise<{ id: string } | null> {
  return apiPost<{ id: string }>("/api/applications", {
    user_id: USER_ID,
    opportunity_id: opportunityId,
    type,
  });
}

File: backend/app/api/private_opportunities.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models.private_opportunity import PrivateOpportunity
from app.schemas.private_opportunity import (
    ParseExternalRequest,
    ParseExternalResponse,
    PrivateOpportunityCreate,
    PrivateOpportunityResponse,
    PrivateOpportunityUpdate,
)
from app.services.parse_external import parse_name, parse_url

router = APIRouter(prefix="/api/opportunities/private", tags=["private-opportunities"])


def _opp_to_response(opp: PrivateOpportunity) -> PrivateOpportunityResponse:
    return PrivateOpportunityResponse(
        id=opp.id,
        user_id=opp.user_id,
        type=opp.type,
        title=opp.title,
        provider=opp.provider,
        description=opp.description,
        eligibility_criteria=opp.eligibility_criteria,
        award_range=opp.award_range,
        deadline=opp.deadline.isoformat() if opp.deadline else None,
        field_tags=opp.field_tags,
        region=opp.region,
        source_url=opp.source_url,
        guidelines=opp.guidelines,
        is_parsed=opp.is_parsed,
        created_at=opp.created_at.isoformat() if opp.created_at else None,
        updated_at=opp.updated_at.isoformat() if opp.updated_at else None,
    )


@router.post("", response_model=PrivateOpportunityResponse, status_code=201)
async def create_private_opportunity(
    payload: PrivateOpportunityCreate,
    db: AsyncSession = Depends(get_db),
):
    if not payload.title or not payload.title.strip():
        raise HTTPException(status_code=422, detail="Title is required")
    if not payload.provider or not payload.provider.strip():
        raise HTTPException(status_code=422, detail="Provider is required")
    if payload.type not in ("grant", "scholarship"):
        raise HTTPException(status_code=422, detail="Type must be 'grant' or 'scholarship'")

    deadline_date = None
    if payload.deadline:
        try:
            deadline_date = date.fromisoformat(payload.deadline)
        except ValueError:
            raise HTTPException(status_code=422, detail="Invalid deadline format. Use YYYY-MM-DD")

    opp = PrivateOpportunity(
        user_id=payload.user_id,
        type=payload.type,
        title=payload.title.strip(),
        provider=payload.provider.strip(),
        description=payload.description,
        eligibility_criteria=payload.eligibility_criteria,
        award_range=payload.award_range,
        deadline=deadline_date,
        field_tags=payload.field_tags,
        region=payload.region,
        source_url=payload.source_url,
        guidelines=payload.guidelines,
        is_parsed=payload.is_parsed,
    )
    db.add(opp)
    await db.commit()
    await db.refresh(opp)
    return _opp_to_response(opp)


@router.get("", response_model=list[PrivateOpportunityResponse])
async def list_private_opportunities(
    user_id: str = Query(..., description="User ID"),
    type: str | None = Query(None, description="Filter by type: grant | scholarship"),
    db: AsyncSession = Depends(get_db),
):
    stmt = select(PrivateOpportunity).where(PrivateOpportunity.user_id == user_id)
    if type:
        stmt = stmt.where(PrivateOpportunity.type == type)
    stmt = stmt.order_by(PrivateOpportunity.created_at.desc())
    result = await db.execute(stmt)
    opps = result.scalars().all()
    return [_opp_to_response(o) for o in opps]


@router.get("/{opp_id}", response_model=PrivateOpportunityResponse)
async def get_private_opportunity(
    opp_id: str,
    db: AsyncSession = Depends(get_db),
):
    opp = await db.get(PrivateOpportunity, opp_id)
    if not opp:
        raise HTTPException(status_code=404, detail="Private opportunity not found")
    return _opp_to_response(opp)


@router.put("/{opp_id}", response_model=PrivateOpportunityResponse)
async def update_private_opportunity(
    opp_id: str,
    payload: PrivateOpportunityUpdate,
    db: AsyncSession = Depends(get_db),
):
    opp = await db.get(PrivateOpportunity, opp_id)
    if not opp:
        raise HTTPException(status_code=404, detail="Private opportunity not found")

    update_data = payload.model_dump(exclude_unset=True)
    if "deadline" in update_data and update_data["deadline"] is not None:
        try:
            opp.deadline = date.fromisoformat(update_data["deadline"])
        except ValueError:
            pass
        del update_data["deadline"]

    for field, value in update_data.items():
        setattr(opp, field, value)

    await db.commit()
    await db.refresh(opp)
    return _opp_to_response(opp)


@router.delete("/{opp_id}", status_code=204)
async def delete_private_opportunity(
    opp_id: str,
    db: AsyncSession = Depends(get_db),
):
    opp = await db.get(PrivateOpportunity, opp_id)
    if not opp:
        raise HTTPException(status_code=404, detail="Private opportunity not found")
    await db.delete(opp)
    await db.commit()


parse_router = APIRouter(prefix="/api/opportunities", tags=["parse-external"])


@parse_router.post("/parse-external", response_model=ParseExternalResponse)
async def parse_external_opportunity(
    payload: ParseExternalRequest,
    db: AsyncSession = Depends(get_db),
):
    if not payload.url and not payload.name:
        raise HTTPException(status_code=400, detail="URL or name is required")

    if payload.name and not payload.url:
        result = await parse_name(payload.name, payload.type)
    else:
        result = await parse_url(payload.url, payload.type)
    return ParseExternalResponse(**result)
